src/read.c

```diff
@@ -4034,137 +3934,99 @@ avifResult avifDecoderNextImage(avifDecoder * decoder)
+            // Nothing changed.
+            continue;
+        }
+        // There is at least one newly decoded tile.
+        assert(group->decodedTileCount > oldDecodedTileCount[alpha]);
 
-            avifImageStealPlanes(decoder->image, srcColor, AVIF_PLANES_YUV);
+        // Grid path.
+        if ((group->grid.rows > 0) && (group->grid.columns > 0)) {
+            assert(group->tileCount == (group->grid.rows * group->grid.columns));
+            AVIF_CHECKERR(avifDecoderDataFillImageGrid(decoder->data, decoder->image, group, oldDecodedTileCount[alpha], alpha),
+                          AVIF_RESULT_INVALID_IMAGE_GRID);
+            continue;
         }
-    }
 
-    if (decoder->data->decodedAlphaTileCount > oldDecodedAlphaTileCount) {
-        // There is at least one newly decoded alpha tile.
-        if ((decoder->data->alphaGrid.rows > 0) && (decoder->data->alphaGrid.columns > 0)) {
-            assert(decoder->data->alphaTileCount == (decoder->data->alphaGrid.rows * decoder->data->alphaGrid.columns));
-            if (!avifDecoderDataFillImageGrid(decoder->data,
-                                              &decoder->data->alphaGrid,
-                                              decoder->image,
-                                              firstAlphaTileIndex,
-                                              oldDecodedAlphaTileCount,
-                                              decoder->data->decodedAlphaTileCount,
-                                              AVIF_TRUE)) {
-                return AVIF_RESULT_INVALID_IMAGE_GRID;
-            }
-        } else {
-            // Normal (most common) non-grid path. Just steal the planes from the only "tile".
-            assert(decoder->data->alphaTileCount == 1);
-            avifImage * srcAlpha = decoder->data->tiles.tile[decoder->data->colorTileCount].image;
-            if ((decoder->image->width != srcAlpha->width) || (decoder->image->height != srcAlpha->height) ||
-                (decoder->image->depth != srcAlpha->depth)) {
+        // Normal (most common) non-grid path.
+        assert(group->tileCount == 1);
+        avifImage * src = decoder->data->tiles.tile[group->firstTileIndex].image;
+        const avifPlanesFlag planes = alpha ? AVIF_PLANES_A : AVIF_PLANES_YUV;
+        if ((decoder->image->width != src->width) || (decoder->image->height != src->height) || (decoder->image->depth != src->depth)) {
+            if (alpha) {
                 avifDiagnosticsPrintf(&decoder->diag, "decoder->image does not match srcAlpha in width, height, or bit depth");
                 return AVIF_RESULT_DECODE_ALPHA_FAILED;
             }
+            avifImageFreePlanes(decoder->image, AVIF_PLANES_ALL);
 
-            avifImageStealPlanes(decoder->image, srcAlpha, AVIF_PLANES_A);
+            decoder->image->width = src->width;
+            decoder->image->height = src->height;
+            decoder->image->depth = src->depth;
+        }
+#if 1
```

Comment:
leftover debug code?

Reply:
I changed the `0` to `1` to verify that the code inside worked properly but forgot to change the `1` back to `0`. Done.